plugins/query-insights/src/main/java/org/opensearch/plugin/insights/rules/model/SearchQueryRecord.java

```diff
@@ -0,0 +1,215 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.plugin.insights.rules.model;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.opensearch.action.search.SearchType;
+import org.opensearch.core.common.io.stream.StreamInput;
+import org.opensearch.core.common.io.stream.StreamOutput;
+import org.opensearch.core.common.io.stream.Writeable;
+import org.opensearch.core.xcontent.ToXContentObject;
+import org.opensearch.core.xcontent.XContentBuilder;
+
+import java.io.IOException;
+import java.util.Locale;
+import java.util.Map;
+
+/**
+ * Simple abstract class that represent record stored in the Query Insight Framework
+ *
+ * @param <T> The value type associated with the record
+ * @opensearch.internal
+ */
+public abstract class SearchQueryRecord<T extends Number & Comparable<T>>
+    implements
+        Comparable<SearchQueryRecord<T>>,
+        Writeable,
```

Comment:
Do we need to implement `Writeable` yet? Right now, we have nothing that requires it.

Reply:
Currently nothing requires it, but I believe there will be cases we want to write the records into a stream output.

Reply:
> but I believe there will be cases we want to write the records into a stream output.

Then we can make the classes writeable then. Never write code because you might need it someday: https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it